lib/pbench/server/database/models/active_tokens.py

```diff
@@ -13,12 +13,6 @@ class ActiveTokens(Database.Base):
     id = Column(Integer, primary_key=True, autoincrement=True)
     token = Column(String(500), unique=True, nullable=False, index=True)
     created = Column(DateTime, nullable=False)
-    user_id = Column(
```

Comment:
I think we'd want to retain the association between "API key" and the user UUID that owns it, in a way that's searchable (against our user cache table) as opposed to having to decrypt the token.

I don't see any reason we need the integer `id` column, however.

Reply:
Do we really want to link the tables together, or should we just store the UUID/userID here?

Reply:
I think it would be handy if we link them together. We can limit the number of tokens the user can have also if we decide to implement different types of tokens then simple join would reveal one-to-many relationship.

Reply:
I think we need two tables (among others...):
- one for translating between OIDC UUIDs and usernames; and,
- one for storing API key tokens

And, I think these tables are distinct from each other.  (Although, you might convince me that they have relationships to each other or to other tables.)

Regarding the translation table, we'll update it every time we're presented with an access token which contains both a UUID and a username.  The UUID should be a unique key; while it would be _nice_ if usernames were unique, I expect that we will have to live with non-unique mappings from username to UUID -- if a given user's username changes, presumably their UUID will _not_, and we would replace the row in the table with the new correspondence.  (Other rows in the table with the same username would have different UUIDs and would be left alone.)  Ownership of datasets would be indicated by UUID; we'll use this table to annotate datasets which are returned to the client by replacing the UUID value with the corresponding username from this table; when we get a query from the client for a particular username, we'll translate it using this table into one or more UUIDs which we'll use to look up the datasets.

Regarding API keys, these have nothing to do with usernames, per se, but we will need to validate them and be able to map them to an owning UUID.  Correspondingly, we'll need to be able to look up all the API keys which are associated with a given UUID, so that the user can maintain them.

However, I don't think we need or want to keep access tokens themselves around (which just need to be able to validate them and pull the UUID and username from them), and I don't think we want to confuse them with access keys (I'm hoping that we will be able to readily tell the two apart).

Reply:
BTW, the `created` time (if we're really going to maintain this at all) should be a `TZDateTime` instead of the default "dezoned" local time.

Reply:
I am not touching `active_tokens` table in this PR as I don't want to step on @siddardh-ra work with API keys, I think when we replace the active_tokens table we will want to have the one (user) to many (tokens) relationship between these two tables.

Reply:
> when we encode the API key we could include the oidc_id in it so that when we decode it we can refer it back, but not addressing that is not a part of this PR.

That would be handy, but it only solves half the problem.  (That would allow the Server to easily get the owner from the key, which is good and important.)  It doesn't allow the owner to find his/her key(s) to manage them.

> We should really always refer to our internal keys as "API Keys" to differentiate them from the OIDC "Access tokens".

❤️ 

> About identifying them apart before decoding, I can only think of the length of the key/token

I don't think that using length is a good approach -- that seems really fragile, as the length could, really, be anything (depending on what is in the key at what times as well as on the details of its encoding).

> I don't see how we can reuse the current oidc validation

Depending on how expensive the decode is, we could just _try_ the OIDC validation, and, if it fails, try our internal validation.  But, we could also do something like require the API key to come from a different HTTP header than the one in which we would look for the access token -- that way, there really is no ambiguity.

Reply:
> Depending on how expensive the decode is, we could just _try_ the OIDC validation, and, if it fails, try our internal validation. But, we could also do something like require the API key to come from a different HTTP header than the one in which we would look for the access token -- that way, there really is no ambiguity.

Yes exactly, the user can specify a different header for the API Keys like `X-API-Key` header and then the server takes a different path for validation.

Reply:
I'm reasonably certain that the requestor knows which type of credential it has:  access tokens come from the dance with the OIDC ID provider (and the client needs to be prepared to renew them), while API keys come from an external source (i.e., they are provided by the user who got them from the Server via some sort of mechanism) and they are stored and reused.

Reply:
I'm thinking of things like the `PbenchServerClient` and `pbench-results-move`, which are generic and ought to be able to work with either an auth token or an API key; and I'd prefer not to have to make the caller tell them (or ideally even need to know).

Reply:
I can see that.  However, I don't think it's going to be practical for a human user to supply an access token to `pbench-results-move` -- I think that that is going to need to be an API key (and so there'll be no abiguity, there).

The `PbenchServerClient` is a slightly different case, but my instinct is that it will likely be access-token-based, so there won't be any ambiguity there, either.  But, to the extent that we want to use API keys for more than just `PUT`, `PbenchServerClient` may need to be enhanced to properly handle each type of credential...whatever that requires.

Reply:
`PbenchServerClient` definitely needs to operate *both* ways, because it's the gateway for functional testing. As for `pbench-results-move`, certainly the most common case is likely to be long-lived API keys, but I'd prefer not to set that restriction in stone if it can be avoided.